[LSV] Merge contiguous chains across scalar types

gandhi56 opened the pull request:
This change enables the LoadStoreVectorizer to merge and vectorize contiguous chains even when their scalar element types differ, as long as the total bitwidth matches. To do so, we rebase offsets between chains, normalize value types to a common integer type, and insert the necessary casts around loads and stores. This uncovers more vectorization opportunities and explains the expected codegen updates across AMDGPU tests.

Key changes:
- Chain merging
  - Build contiguous subchains and then merge adjacent ones when:
    - They refer to the same underlying pointer object and address space.
    - They are either all loads or all stores.
    - A constant leader-to-leader delta exists.
    - Rebasing one chain into the other's coordinate space does not overlap.
    - All elements have equal total bit width.
  - Rebase the second chain by the computed delta and append it to the first.

- Type normalization and casting
  - Normalize merged chains to a common integer type sized to the total bits.
  - For loads: create a new load of the normalized type, copy metadata, and cast back to the original type for uses if needed.
  - For stores: bitcast the value to the normalized type and store that.
  - Insert zext/trunc for integer size changes; use bit-or-pointer casts when sizes match.

- Cleanups
  - Erase replaced instructions and DCE pointer operands when safe.
  - New helpers: computeLeaderDelta, chainsOverlapAfterRebase, rebaseChain, normalizeChainToType, and allElemsMatchTotalBits.

Impact:
- Increases vectorization opportunities across mixed-typed but size-compatible access chains.
- Large set of expected AMDGPU codegen diffs due to more/changed vectorization.

This PR resolves #97715.

arsenm commented on llvm/lib/Transforms/Utils/Local.cpp:
```diff
@@ -3496,6 +3496,51 @@ void llvm::copyMetadataForLoad(LoadInst &Dest, const LoadInst &Source) {
+      break;
+
+    case LLVMContext::MD_tbaa: {
+      MDNode *NewTyNode =
+          MDB.createTBAAScalarTypeNode(NewType->getStructName(), N);
+      Dest.setMetadata(LLVMContext::MD_tbaa, NewTyNode);
```
I'm not sure you can safely preserve this in the vectorization case

gandhi56 replied:
Can you please elaborate? Vectorization hasn't occurred yet, and this function should serve a general scenario where metadata is being copied over.

arsenm commented on llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp:
```diff
@@ -424,6 +432,20 @@ PreservedAnalyses LoadStoreVectorizerPass::run(Function &F,
   return Changed ? PA : PreservedAnalyses::all();
 }
 
+static const Value *getUnderlyingPtrObject(const Value *Ptr) {
+  const Value *ObjPtr = llvm::getUnderlyingObject(Ptr);
+  if (const auto *Sel = dyn_cast<SelectInst>(ObjPtr)) {
+    // The select's themselves are distinct instructions even if they share
+    // the same condition and evaluate to consecutive pointers for true and
+    // false values of the condition. Therefore using the select's themselves
+    // for grouping instructions would put consecutive accesses into different
+    // lists and they won't be even checked for being consecutive, and won't
+    // be vectorized.
+    return Sel->getCondition();
+  }
+  return ObjPtr;
+}
```
I don't understand this function, and don't really follow the comment. 

"PtrObject" sounds like it's supposed to return a pointer to an object, which it does, unless it's a select, in which case it returns a boolean value? Plus what happens if the pointer is a select of 2 pointers derived from the same base object?

arsenm commented on llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp:
```diff
@@ -467,6 +489,98 @@ bool Vectorizer::run() {
+          Builder.CreateStore(CastVal, SI->getPointerOperand());
+      NewStore->setAlignment(SI->getAlign());
+      NewStore->setVolatile(SI->isVolatile());
+      copyMetadataForStore(*NewStore, *SI);
```
This is creating an equivalently sized store, that hasn't been vectorized?

arsenm commented on llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp:
```diff
@@ -479,49 +593,111 @@ bool Vectorizer::runOnPseudoBB(BasicBlock::iterator Begin,
+
+    // Normalize the value operand/result type of each instruction in C1 to
+    // C1CastTy.
+    Type *C1CastTy =
+        Type::getIntNTy(C1ElemTy->getContext(), DL.getTypeSizeInBits(C1ElemTy));
```
Slightly nicer to directly use IntegerType

gandhi56 committed:
[LSV] Merge contiguous chains across scalar types

This change enables the Load/Store Vectorizer to
merge and vectorize contiguous chains even when
their scalar element types differ, as long as the
total bitwidth matches. To do so, we rebase
offsets between chains, normalize value types to
a common integer type, and insert the necessary
casts around loads and stores. This uncovers more
vectorization opportunities and explains the
expected codegen updates across AMDGPU tests.

Key changes:
- Chain merging
  - Build contiguous subchains and then merge
  adjacent ones when:
    - They refer to the same underlying pointer
    object and address space.
    - They are either all loads or all stores.
    - A constant leader-to-leader delta exists.
    - Rebasing one chain into the other's
    coordinate space does not overlap.
    - All elements have equal total bit width.
  - Rebase the second chain by the computed delta
  and append it to the first.

- Type normalization and casting
  - Normalize merged chains to a common integer
  type sized to the total bits.
  - For loads: create a new load of the normalized
  type, copy metadata, and cast back to the
  original type for uses if needed.
  - For stores: bitcast the value to the
  normalized type and store that.
  - Insert zext/trunc for integer size changes;
  use bit-or-pointer casts when sizes match.

- Cleanups
  - Erase replaced instructions and DCE pointer
  operands when safe.
  - New helpers: computeLeaderDelta,
  chainsOverlapAfterRebase, rebaseChain,
    normalizeChainToType, and
    allElemsMatchTotalBits.

Impact:
- Increases vectorization opportunities across
mixed-typed but size-compatible access chains.
- Large set of expected AMDGPU codegen diffs due
to more/changed vectorization.

Add more tests

Refactor metadata copying for loads and stores

Add tests for metadata copy after load/store merging

gandhi56 commented on llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp:
```diff
@@ -467,6 +489,98 @@ bool Vectorizer::run() {
 
+Value *Vectorizer::insertCast(Value *Val, Type *DstTy) {
+  if (DL.getTypeSizeInBits(Val->getType()) == DL.getTypeSizeInBits(DstTy)) {
+    return Builder.CreateBitOrPointerCast(Val, DstTy, Val->getName() + ".bc");
```
`CreateBitOrPointerCast` does not emit an addrspacecast. It's `CreatePointerBitCastOrAddrSpaceCast` which does.

gandhi56 commented on llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp:
```diff
@@ -479,49 +593,111 @@ bool Vectorizer::runOnPseudoBB(BasicBlock::iterator Begin,
+    auto *PtrOperand = getLoadStorePointerOperand(I);
+    if (I->use_empty())
+      I->eraseFromParent();
+    RecursivelyDeleteTriviallyDeadInstructions(PtrOperand);
```
No, the operand needs to be handled separately.

gandhi56 commented:
Internal PSDB passed.

LU-JOHN commented on llvm/lib/Transforms/Utils/Local.cpp:
```diff
@@ -3140,6 +3140,51 @@ void llvm::copyMetadataForLoad(LoadInst &Dest, const LoadInst &Source) {
+        Dest.setMetadata(ID, N);
+      break;
+
+    case LLVMContext::MD_range:
```
Range metadata is not valid for stores.  https://llvm.org/docs/LangRef.html#range-metadata

gandhi56 commented on llvm/lib/Transforms/Utils/Local.cpp:
```diff
@@ -3496,6 +3496,51 @@ void llvm::copyMetadataForLoad(LoadInst &Dest, const LoadInst &Source) {
+      Dest.setMetadata(LLVMContext::MD_tbaa, NewTyNode);
+      break;
+    }
+    case LLVMContext::MD_nonnull:
```
How about setting it if the value operand of the new store is a pointer?

gandhi56 replied:
StoreInst.getType() would yield `VoidTy` so it probably doesn't make sense to set `MD_nonnull`.

nikic reviewed (commented):
I've not looked at the LSV changes (leaving this to @arsenm), but I'll note that the IR test coverage here looks pretty weak for a significant change to the pass. The part I looked at (the Local.cpp changes) don't appear to have any test coverage at all.

gandhi56 commented:
I added a comprehensive set of normalization tests in merge-vectors-complex.ll, functionality tests in merge-vectors.ll to ensure all checks are doing their job before deciding to merge loads/stores and finally introduced copy-metadata-loads-stores.ll for metadata preservation checks.

gandhi56 committed:
Add metadata-copying tests

gandhi56 committed:
Rebase, update tests

gandhi56 commented on llvm/lib/Transforms/Utils/Local.cpp:
```diff
@@ -3153,6 +3153,48 @@ void llvm::copyMetadataForLoad(LoadInst &Dest, const LoadInst &Source) {
+
+    case LLVMContext::MD_tbaa: {
+      MDNode *NewTyNode =
+          MDB.createTBAAScalarTypeNode(NewType->getStructName(), N);
```
Removed the store block.

dakersnar commented on llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp:
```diff
@@ -1625,8 +1782,7 @@ std::vector<Chain> Vectorizer::gatherChains(ArrayRef<Instruction *> Instrs) {
   for (auto &E : MRU)
-    if (E.second.size() > 1)
-      Ret.emplace_back(std::move(E.second));
+    Ret.emplace_back(std::move(E.second));
```
Can you explain this change? Are you basically allowing chains that have a size of 1 in order to merge them later?

dakersnar commented on llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp:
```diff
@@ -480,49 +582,120 @@ bool Vectorizer::runOnPseudoBB(BasicBlock::iterator Begin,
-    Changed |= runOnChain(C);
-  return Changed;
-}
+  // Merge chains in reverse order, so that the first chain is the largest.
```
This comment concerns me. We don't have a guarantee that chains are in increasing size order at this point in the code. Am I missing something?

dakersnar commented on llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp:
```diff
@@ -480,49 +582,120 @@ bool Vectorizer::runOnPseudoBB(BasicBlock::iterator Begin,
+               C1.back().OffsetFromLeader + *DeltaOpt;
+    if (!Sz.isPowerOf2())
+      continue;
+
```
Do you mind explaining this calculation? This doesn't seem right to me.
